contracts/extensions/TokenSupplier.sol

```diff
@@ -0,0 +1,191 @@
+
+    tokenSupplyCeiling = _tokenSupplyCeiling;
+    tokenIssuanceRate = _tokenIssuanceRate;
+    lastPinged = block.timestamp;
+    lastRateUpdate = block.timestamp;
+  }
+
+  /// @notice Update the tokenSupplyCeiling, cannot set below current tokenSupply
+  /// @param _tokenSupplyCeiling Total amount of tokens to issue
+  function setTokenSupplyCeiling(uint256 _tokenSupplyCeiling) public initialised onlyRoot onlyContract {
+    try ColonyExtension(msg.sender).identifier() returns (bytes32 extensionId) {
+      IColonyNetwork colonyNetwork = IColonyNetwork(colony.getColonyNetwork());
+      address installation = colonyNetwork.getExtensionInstallation(extensionId, address(colony));
+      require(installation == msg.sender, "token-supplier-not-managed-extension");
+      require(extensionId == VOTING_HYBRID, "token-supplier-cannot-set-value");
+    } catch {
+      require(false, "token-supplier-no-identifier");
+    }
+
+    tokenSupplyCeiling = _tokenSupplyCeiling;
+
+    emit TokenSupplyCeilingSet(tokenSupplyCeiling);
+  }
+
+  /// @notice Update the tokenIssuanceRate
+  /// @param _tokenIssuanceRate Number of tokens to issue per day
+  function setTokenIssuanceRate(uint256 _tokenIssuanceRate) public initialised onlyRoot onlyContract {
+    try ColonyExtension(msg.sender).identifier() returns (bytes32 extensionId) {
+      IColonyNetwork colonyNetwork = IColonyNetwork(colony.getColonyNetwork());
+      address installation = colonyNetwork.getExtensionInstallation(extensionId, address(colony));
+      require(installation == msg.sender, "token-supplier-not-managed-extension");
+      require(
+        extensionId == VOTING_HYBRID || (
+          extensionId == VOTING_REPUTATION &&
+          block.timestamp - lastRateUpdate >= 4 weeks &&
+          _tokenIssuanceRate <= add(tokenIssuanceRate, tokenIssuanceRate / 10) &&
+          _tokenIssuanceRate >= sub(tokenIssuanceRate, tokenIssuanceRate / 10)
+        ),
+        "token-supplier-cannot-set-value"
+      );
+    } catch {
+      require(false, "token-supplier-no-identifier");
+    }
+
+    // Issue any outstanding tokens under the previous rate
+    issueTokens();
+
+    tokenIssuanceRate = _tokenIssuanceRate;
+    lastRateUpdate = block.timestamp;
+
+    emit TokenIssuanceRateSet(tokenIssuanceRate);
+  }
+
+  /// @notice Issue the appropriate amount of tokens
+  function issueTokens() public initialised {
+    address token = colony.getToken();
+    uint256 tokenSupply = ERC20Extended(token).totalSupply();
+
+    uint256 newSupply = min(
+      sub(tokenSupplyCeiling, tokenSupply),
+      wmul(tokenIssuanceRate, wdiv((block.timestamp - lastPinged), ISSUANCE_PERIOD))
+    );
+
+    assert(add(tokenSupply, newSupply) <= tokenSupplyCeiling);
+
+    // Don't update lastPinged if we aren't actually issuing tokens
+    if (newSupply > 0) {
+      colony.mintTokens(newSupply);
+      lastPinged = block.timestamp;
+    }
+
+    emit TokensIssued(newSupply);
```

Comment:
This function not throwing if there are no tokens to issue is fine and probably useful in places, but I think this event should only be emitted if `newSupply > 0` for client spam reasons.